doc/logbook.1
.TH LOGBOOK 1
.SH NAME
logbook \- a tool for keeping a personal log
.SH SYNOPSIS
.B logbook
[\fB\-\-file\fR \fIPATH\fR]
[\fB\-\-public\fR | \fB\-\-semi\-private\fR | \fB\-\-private\fR]
[\fB\-\-markdown\fR | \fB\-\-paragraph\fR]
[\fB\-\-template \fIPATH\fR]
[\fB\-\-title \fISTRING\fR]
.SH DESCRIPTION
.B logbook
generates a HTML page from a
.BR logbook-log (7)
file including its content based on given privacy settings and prints it to standard output.
.SH OPTIONS
.TP
.B \-\-file \fIPATH\fR
Use \fIPATH\fR as input log file instead of standard input.
.TP
.B \-\-public
Sets the privacy level to public meaning that only log items with this privacy level
(signified using a leading \fB+\fR) will be included in the output HTML. Should not be given
at the same time as \fB\-\-semi-private\fR or \fB\-\-private\fR (the last one given will
overwrite the previous ones).
.TP
.B \-\-semi-private
Sets the privacy level to semi-private meaning that only semi-private and public log items
will be included in the output HTML (signified using \fB*\fR and \fB+\fR respectively).
The semi-private level is typically used when creating a rendering of a log file for
trusted individuals. This option should not be given at same time as \fB\-\-public\fR or
\fB\-\-private\fR (the last one given will overwrite the previous ones).
.TP
.B \-\-private
Sets the privacy level to private which means that all log items will be included in the
output HTML. This level is typically used for the author(s) of the log file.
Private log items are signified using \fB+\fR. This option should not be given at same
time as \fB\-\-public\fR or \fB\-\-semi-private\fR (the last one given will overwrite
the previous ones).
.TP
.B \-\-markdown
Enables markdown markup. When this is enabled, \fBlogbook\fR will convert any text in the
log file to HTML using markdown. This happens individually for every text block.
.TP
.B \-\-paragraph
Enables a "fake markup" which passes trough the text without processing it in any way
except for wrapping it with \fB<p>\fI...\fB</p>\fR. This emulates a behavior of markdown
which also wraps normal text blocks with an HTML paragraph and is intended to make
templates written with markdown markup in mind work better with plain text log files.
This was the default behavior of \fBlogbook\fR 0.2 and prior, so \fB\-\-paragraph\fR
can be used to restore the old behavior. Should not be set at the same time as
\fB\-\-markdown\fR (the option set last will overwrite the previous one).
.TP
.B \-\-template \fIPATH\fR
Read the Jingoo template to be used for HTML generation from \fIPATH\fR instead of using
the default one. Jingoo is a template engine for OCaml which is almost compatible with
Python's Jinja2. For more information read
.BR logbook-template (7).
.TP
.B \-\-title \fISTRING\fR
Set the \fBtitle\fR variable for the template to \fISTRING\fR. In the default template
and in typical cases it is used to set the content of the HTML page's \fB<title>\fR tag.
.SH EXAMPLES
Read \fBmy.log\fR and generate HTML page with public privacy level:
.PP
.nf
.RS
logbook \-\-file my.log \-\-public
.RE
.fi
.PP
Write a markdown rendering of \fBmy.log\fR to \fBfriends.html\fR with privacy
level semi private titled "my log".
.PP
.nf
.RS
logbook \-\-file my.log \-\-semi-private \-\-title "my log" \-\-markdown > friends.html
.RE
.fi
.PP
Read a remote log file using ssh and generate a private privacy level rendering of it
using a custom template
.PP
.nf
.RS
ssh foo@bar "cat /path/to/my.log" | logbook --private --template custom.jingoo
.RE
.fi
.PP
Generate an HTML output for the \fBlogbook\fR changelog:
.PP
.nf
.RS
logbook --file /path/to/doc/logbook/CHANGES.log --markdown > CHANGES.html
.RE
.fi
.SH SEE ALSO
.BR logbook-log (7),
.BR logbook-template (7)


CHANGES.md


v0.2 2020-4-24
------------------------------------

* Switched to dune:
  Use the new (quasi) standard OCaml build system.
* Fix build errors:
  Mostly related to dependencies updating and breaking their API.

v0.1 2017-2-18
------------------------------------

* Initial Version:
  Never really released.





src/logbook.ml
open Stdlib
open Lwt.Infix
open Cow
open Jingoo

let parse_file f =
  let file_parser c =
    Lwt_io.read c >>= (fun s ->
      Lwt.return (Result.map_error (fun e -> "Parse error:" ^ e)
                   (Angstrom.parse_string ~consume:All Log.log_parser s)))
  in match f with
     | None -> file_parser Lwt_io.stdin
     | Some filename ->
         try (Lwt_io.with_file ~mode:Lwt_io.Input filename file_parser)
         with Unix.Unix_error (e, _, _) ->
           Lwt.return (Result.Error ("Failed opening file:" ^ Unix.error_message e))

let get_template f =
  match f with
  | None -> Result.Ok Logbook_template.template
  | Some name ->
      try Result.Ok (Lwt_main.run (Lwt_io.with_file ~mode:Lwt_io.Input name Lwt_io.read))
      with Unix.Unix_error (e, _, _) ->
        Result.Error ("Failed reading template:" ^ Unix.error_message e)

let html_of_log log markup template privacy title =
  try Result.Ok (Jg_template.from_string template
        ~models:(Logbook_models.model_of_log title privacy
          (Log.apply_markup (fun
            x -> Xml.to_string ~decl:false (markup x)) log)))
     with Jg_types.SyntaxError msg ->
       Result.Error ("Template syntax error:" ^ msg)

let input_file = ref None
let privacy = ref Log.Public
let markup = ref (fun s -> Html.string s)
let title = ref "log"
let template_file = ref None

let arglist =
  [ ("--file", Arg.String (fun f -> input_file := Some f), "log file to use");
    ("--private", Arg.Unit (fun () -> privacy := Log.Private),
    "set privacy level of output to private");
    ("--public", Arg.Unit (fun () -> privacy := Log.Public),
    "set privacy level of output to public");
    ("--semi-private", Arg.Unit (fun () -> privacy := Log.Semi_private),
    "set privacy level of output to semi-private");
    ("--markdown", Arg.Unit (fun () -> markup := Markdown.of_string),
    "enable markdown markup");
    ("--paragraph", Arg.Unit (fun () -> markup := (fun s -> Html.p (Html.string s))), "Wrap summary, item titles and texts with <p>, like for example markdown does, but without applying any markup");
    ("--title", Arg.String (fun f -> title := f), "title of the generated html document");
    ("--template", Arg.String (fun f -> template_file := Some f), "Jingoo template to use for HTML generation");
  ]

let usage =
  Sys.argv.(0) ^ " [--file <file.log>]" ^
  " [--template <template.html.jingoo>]" ^
  " [--public | --private | --semi-private]" ^
  " [--markdown]" ^ " [--title <title>]"

let _ =
  Arg.parse arglist (fun _ -> ()) usage;
  let result = Result.bind (Lwt_main.run (parse_file !input_file)) (fun log ->
    Result.bind (get_template !template_file) (fun tpl ->
      html_of_log log !markup tpl !privacy !title))
  in match result with
     | Result.Ok s -> print_endline s; exit 0
     | Result.Error e -> print_endline e; exit 1
